docs: add "Creating Composer Packages" (#7742)
codeigniter4/CodeIgniter4: user_guide_src/source/extending/composer_packages.rst

##########################
Creating Composer Packages
##########################

You can make the :doc:`../general/modules` you create into Composer packages,
or can create a Composer package for CodeIgniter 4.

.. contents::
    :local:
    :depth: 2

****************
Folder Structure
****************

Here's a typical directory structure for a Composer package::

    your-package-name/
    ├── README.md
    ├── composer.json
    ├── src/
    │   └── YourClass.php
    └── tests/

>         └── YourClassTest.php

Reviewer: I think `.gitignore` can be added so `vendor/` can be excluded so it "publish-ready" to packagist:

```
# .gitignore
/vendor
```

Author: Added. Thanks.

##########################
Creating Composer Packages
##########################

You can make the :doc:`../general/modules` you create into Composer packages,
or can create a Composer package for CodeIgniter 4.

.. contents::
    :local:
    :depth: 2

****************
Folder Structure
****************

Here's a typical directory structure for a Composer package::

>     your-package-name/

Reviewer 2: If it would be published to packagist, then it should have a `LICENSE`.

Author: Good point. Added.

  such as the name of you or your organization.
- ``your-package-name``: The name of the package you are creating.

Thus, it is important to make the name unique and distinguish it from other packages.
Uniqueness is especially important when publishing.

Namespace
=========

The package name then determines the vendor namespace in ``autoload.psr4``.

If your package name is ``your-vendor/your-package``, the vendor namespace must
be ``YourVendor``. So you would write like the following::

    "autoload": {
        "psr-4": {
            "YourVendor\\YourPackage\\": "src/"
        }
    },

This setting instructs Composer to autoload the source code for your package.

***************************
Preparing Development Tools
***************************

There are many tools that help ensure quality code. So you should use them.
You can easily install and configure such tools with
`CodeIgniter DevKit <https://github.com/codeigniter4/devkit>`_.

Installing DevKit
=================

In the root of your package directory, run the following commands::

    > composer config minimum-stability dev
    > composer config prefer-stable true

>     > composer require --dev codeigniter4/devkit

Reviewer 2: The `>` can be removed so that the commands can be copy pasted.

Author: We now use `>` in all command samples in the user guide.
Removing it only here is not consistent.

Author: I sent #7757 and fixed these.

Creating composer.json
**********************

In the root of your package directory, create a **composer.json** file. This file
defines metadata about your package and its dependencies.

The ``composer init`` command helps you create it.

For example, **composer.json** might look like this::

    {
        "name": "your-vendor/your-package",
        "description": "Your package description",
        "type": "library",
        "license": "MIT",
        "autoload": {
            "psr-4": {
                "YourVendor\\YourPackage\\": "src/"
            }
        },
        "authors": [
            {
                "name": "Your Name",
                "email": "yourname@example.com"
            }
        ],
        "require": {
            // Any dependencies required by your package go here
        },
        "require-dev": {
            // Any development dependencies (e.g., PHPUnit) go here
        }
    }

Package Name
============

The ``name`` field is important here. Package names are generally written in the
format "vendor-name/package-name" with all lowercase. Here is a common example:

- ``your-vendor-name``: The name that identifies the vendor (creator of the package),

>   such as the name of you or your organization.

Reviewer 2: ```suggestion
  such as your name or your organization.
```

- ``your-package-name``: The name of the package you are creating.

> Thus, it is important to make the name unique and distinguish it from other packages.

Reviewer 2: ```suggestion
Thus, it is important to make the name unique to distinguish it from other packages.
```

Uniqueness is especially important when publishing.

Namespace
=========

The package name then determines the vendor namespace in ``autoload.psr4``.

> If your package name is ``your-vendor/your-package``, the vendor namespace must

Reviewer 2: Not always true. I see packages where it is all lowercase. Some removes some letters.
```suggestion
If your package name is ``your-vendor/your-package``, the vendor namespace will typically
```

Author: > I see packages where it is all lowercase. 

I see. It could be. PHP's classnames are case insensitive.

> Some removes some letters.

Why? Isn't it just a incorrect use, or just a typo?
If it is not the same, the vendor namespace might conflict.

Reviewer 2: like in `pestphp/pest`, the vendor namespace is Pest.

Author: Isn't it just a incorrect use?
If someone gets the vendor name `pest` on packagist, they would use the vendor name space `Pest`?

Author: If we follow the PSR, the class name is CamelCase. So I think the current wording is fine.
Or not `must` but `should`?

Reviewer 2: Let's leave it as is for now.

Configuring Coding Standards Fixer
==================================

DevKit provides Coding Standards Fixer with
`CodeIgniter Coding Standard <https://github.com/CodeIgniter/coding-standard>`_
based on `PHP-CS-Fixer <https://github.com/PHP-CS-Fixer/PHP-CS-Fixer>`_.

Copy **vendor/codeigniter4/devkit/src/Template/.php-cs-fixer.dist.php** to your
project root folder.

Create the **build** folder for the cache file::

    your-package-name/
    ├── .php-cs-fixer.dist.php
    ├── build/

Open **.php-cs-fixer.dist.php** in your editor, and fix the folder path::

    --- a/.php-cs-fixer.dist.php
    +++ b/.php-cs-fixer.dist.php
    @@ -7,7 +7,7 @@ use PhpCsFixer\Finder;
     $finder = Finder::create()
         ->files()
         ->in([
    -        __DIR__ . '/app/',
    +        __DIR__ . '/src/',
             __DIR__ . '/tests/',
         ])
         ->exclude([

That't it. Now you can run Coding Standards Fixer::

    > vendor/bin/php-cs-fixer fix --ansi --verbose --diff

If you add ``scripts.cs-fix`` in your **composer.json**, you can run it with
``composer cs-fix`` command::

    --- a/composer.json
    +++ b/composer.json
    @@ -23,5 +23,8 @@
             "allow-plugins": {

>                  "phpstan/extension-installer": true

Reviewer 2: this part can be removed as this section talks about cs

Author: Changed the sample code.
